crates/storage/nippy-jar/src/writer.rs

```
@@ -0,0 +1,341 @@
+                    // Represents the offset of the soon to be appended data column
+                    self.offsets.push(self.data_file.stream_position()?);
+                }
+
+                self.write_column(value.as_ref())?;
+
+                // Last offset represents the size of the data file if no more data is to be
+                // appended. Otherwise, represents the offset of the next data item.
+                self.offsets.push(self.data_file.stream_position()?);
+            }
+            None => {
+                return Err(NippyJarError::UnexpectedMissingValue(
+                    self.jar.rows as u64,
+                    self.column as u64,
+                ))
+            }
+            Some(Err(err)) => return Err(err.into()),
+        }
+
+        Ok(())
+    }
+
+    /// Writes column to data file. If it's the last column of the row, call `finalize_row()`
+    fn write_column(&mut self, value: &[u8]) -> Result<(), NippyJarError> {
+        self.uncompressed_row_size += value.len();
+        if let Some(compression) = &self.jar.compressor {
+            let before = self.tmp_buf.len();
+            let len = compression.compress_to(value, &mut self.tmp_buf)?;
+            self.data_file.write_all(&self.tmp_buf[before..before + len])?;
+        } else {
+            self.data_file.write_all(value)?;
+        }
+
+        self.column += 1;
+
+        if self.jar.columns == self.column {
+            self.finalize_row();
+        }
+
+        Ok(())
+    }
+
+    /// Prunes rows from data and offsets file and updates its configuration on disk
+    pub fn prune_rows(&mut self, num_rows: usize) -> Result<(), NippyJarError> {
+        // Each column of a row is one offset
+        let num_offsets = num_rows * self.jar.columns;
+
+        // Calculate the number of offsets to prune from in-memory list
+        let offsets_prune_count = num_offsets.min(self.offsets.len().saturating_sub(1)); // last element is the expected size of the data file
```

should we return an error in case when `num_offsets >= self.offsets`, rather than calculating the `min` and proceeding with it?

---

not, because this is just regarding the in memory offset list. which is the one not yet flushed to disk (for whatever reason).

but overall (taking into account offsets in disk), i was just ignoring that case.... but i can see how it can be useful